packages/i/iksemel/files/0001-Escape-non-ASCII-characters.patch

```diff
@@ -0,0 +1,68 @@
+From b0073153eaa6da67ce9d0674f47a25b8f9dce803 Mon Sep 17 00:00:00 2001
+From: Silke Hofstra <silke@slxh.eu>
+Date: Tue, 2 Apr 2024 22:13:31 +0200
+Subject: [PATCH] Escape non-ASCII characters
+
+Escape characters that are not contained in ASCII by encoding them to their HTML/Hex representation.
+---
+ src/iks.c | 23 +++++++++++++++++++++++
+ 1 file changed, 23 insertions(+)
+
+diff --git a/src/iks.c b/src/iks.c
+index d1a1a93..d517686 100644
+--- a/src/iks.c
++++ b/src/iks.c
+@@ -4,6 +4,7 @@
+ ** modify it under the terms of GNU Lesser General Public License.
+ */
+ 
++#include <uchar.h>
+ #include "common.h"
+ #include "iksemel.h"
+ 
+@@ -569,10 +570,18 @@ escape_size (char *src, size_t len)
+ 	size_t sz;
+ 	char c;
+ 	int i;
++	char32_t mb;
++	mbstate_t state = {0};
+ 
+ 	sz = 0;
+ 	for (i = 0; i < len; i++) {
+ 		c = src[i];
++
++		int rc = (int)mbrtoc32(&mb, src+i, len - i, &state);
++		if (rc > 1) {
++			sz += rc + 3;
++		}
++
+ 		switch (c) {
+ 			case '&': sz += 5; break;
+ 			case '\'': sz += 6; break;
+@@ -599,9 +608,23 @@ escape (char *dest, char *src, size_t len)
+ 	char c;
+ 	int i;
+ 	int j = 0;
++	char32_t mb;
++	char buf[13] = {0};
++	mbstate_t state = {0};
+ 
+ 	for (i = 0; i < len; i++) {
+ 		c = src[i];
++
++		int rc = (int)mbrtoc32(&mb, src+i, len - i, &state);
++		if (rc < 0 || rc > 1) {
```

Comment:
I'm a little worried about this. So, `< 0` is error, but `0` is also returned for NULL input. The check would appear to deal with certain errors, ignore others.. ?

Reply:
Yes. This essentially skips invalid and continuation bytes in the sequence. 

There are three documented error cases for `mbrtoc32`, two of which rely on mult-`char32_t ` sequences (should not be possible within unicode). The third error case happens when an invalid character is encountered, which happens in two situations:

- A continuation byte: should be skipped as it is already encoded.
- An actually invalid sequence: should be skipped in order to produce valid UTF-8 output.

I have added a unit test for both scenarios.